Feature/ingredient class update.jw

jwasmer opened the pull request:
What does this PR do?
- Modifies the Ingredient class to contain a static method that constructs itself. The data for its construction is passed in as a parameter, ingredientSource, along with our array of all ingredient objects data.

Where should your reviewer start?
- Start by reviewing the Ingredient.js file (almost the entire file has been changed), from there move on to Recipe.js and review the changes made to the getAllIngredientData function. The remaining changes are changes to the test suite.

What (if any) features are you implementing?
- Static method to build Ingredient instance.

What (if anything) did you refactor?
- Refactored Ingredient.js and Recipe.js getAllIngredientData().

Were there any issues that arose?
- Some changes to Ingredient-test.js were necessary to make the test suite compatible with the new static method.

Is there anything that you need from your teammate?

Any other comments, questions, or concerns to add?

jwasmer committed:
Build static method to instantiate Ingredient class from available data

jwasmer committed:
Merge branch 'main' of github.com:grainymac/whatsCookin

jwasmer committed:
Update the Ingredient class with a static method to create a new Ingredient instance

jwasmer committed:
Update Recipe getAllIngredientsData method to use new Ingredient static method

jwasmer committed:
Update Ingredient test files

jwasmer committed:
Update Recipe test files

jwasmer commented:
The intent behind this change was to allow the `Pantry` data to be pulled by calling the `Ingredient` static method directly rather than having to piece separate data together. Creating a new `Ingredient` class instance just requires executing the static method and passing through the ingredient ID and an array of all ingredient data. If `Recipe` data is available that can be added as a third parameter but it is not required. This makes the process of instantiating new `Ingredient` classes more flexible for us down the road.

tricia-holmes commented on src/classes/Ingredient.js:
```diff
@@ -1,11 +1,30 @@
+    this.unit = data?.unit
+  }
+
+  static fromIngredientData(id, ingredientsData, recipeData = null) {
+    const fromIngredients = ingredientsData.find(ingredient => id === ingredient.id)
+    let fromRecipe;
+    
+    let allIngredientData = {}
+
+    allIngredientData.id = id
+    allIngredientData.name = fromIngredients.name
+    allIngredientData.cost = fromIngredients.estimatedCostInCents
```
Was this also being refactored to keep whole pantry ingredients? If it's a pantry ingredient, it would also need to have an `amount` set for it. Not sure if this was just for ingredients and recipe ingredients, or if we were transforming all ingredients (`ingredient`, `pantry ingredient`, `recipe ingredient` to be an instance of ingredient class.

tricia-holmes commented on src/classes/Ingredient.js:
```diff
@@ -1,11 +1,30 @@
+    this.id = data.id
+    this.name = data.name
+    this.cost = data.cost
+    this.amount = data?.amount
+    this.unit = data?.unit
```
Hey! I believe `optional chaining` is for nested object properties, and since `amount` and `unit` are just properties they will already default to `undefined`. I tried without it and console logged in the test the ingredient that didn't have `amount` and `unit` given. The result is `undefined`. Here is a screenshot below:

![image](https://user-images.githubusercontent.com/108706408/199750828-ca6a00fd-8255-431e-9fc4-5e45fdbd6416.png)

jwasmer replied:
Sounds good! I can push up those changes.

jwasmer committed:
Refactor static method to accept two parameters, ingredientSource and ingredientsData

jwasmer committed:
Update getAllIngredientsData to accept only ingredientsData as a parameter

jwasmer committed:
Refactor Ingredient tests to reflect changes to static method

jwasmer committed:
Remove an unnecessary chain operator

jwasmer commented:
New changes. `fromIngredientData` only accepts two parameters now, `ingredientSource` and `ingredientsData`. 

`ingredientSource` is the data that's prompting the class to be created. This should either be coming from `Recipe` for the `getAllIngredientsData` method, or from the `pantry` in the `User` class.

`ingredientsData` is the array of all ingredients data.

Got rid of the unnecessary chaining, updated the tests (again), and updated `getAllIngredientsData` to get rid of the now unnecessary `recipeData` parameter.

Adelle-Pitsas commented:
Pulled it down, and it looks good to me!

tricia-holmes commented:
Looks good! We will need to refactor `User` to now invoke the static method for ingredients, but that shouldn't be a problem. Merging in this PR!
